### ISSUE TITLE
Filter component focus lost

### ISSUE THREAD

**Issue by DrunkenMoose** [fooplugins/FooTable#602]
Hey,

First of all thanks for this cool plugin, it works very well. I have one question about something i ran into.
I enabled the filter functionality and when the user is done typing the focus is lost on the filtering inputbox. Is there a way to still have this focus on the inputbox? I think this is more user friendly.

Greetings from the Netherlands,
Ro

**Comment by jeffconklin** [comment]
@DrunkenMoose  - I 100% agree this would be a good addition. I have gotten into the habit of using the tab/shift+tab and arrows keys to get focus again. If it just kept the focus where ever your cursor was, that would be much better.

**Comment by steveush** [comment]
Hi,

This has been resolved in the develop branch and will be included in the next release.

Thanks

**Comment by givemetraffic** [comment]
Hi 

@steveush  Sorry but, this is the worst fix ever. Instead of keep input from loosing it's focus you are just reapplying focus after every filter run. Try to use it with 300ms delay. Did you ever seen how other people implement autocomplete or filters?

Moreover it's default behavior now. Ridiculous.

Just keep filter form outside the table, it's so simple
```
self.$row = $('<tr/>', {'class': 'footable-filtering'}).prependTo(self.ft.$el.children('thead'));
self.$cell = $('<th/>').attr('colspan', self.ft.columns.visibleColspan).appendTo(self.$row);
self.$form = $('<form/>', {'class': 'form-inline'}).append($form_grp).appendTo(self.$cell);
```
To
```
self.$form = $('<form/>', {'class': 'form-inline'}).append($form_grp).insertBefore(self.ft.$el);
```

And fix styles

**Comment by jeffconklin** [comment]
I have not tried the fix in the dev branch, but if it is anything other than the focus not being lost at all, this is a tad frustrating. I use filtering hundreds of times a day most days and this bug has been frustrating for some time. If my cursor is not where I "left it" after the filter is applied, I agree with @givemetraffic that this is still an issue and a pretty big flaw.

@steveush , I know that you have implemented a feature where we can place the filter anywhere.. like outside the table as giveme mentioned. If we place it outside the table will is still have an issue?

**Comment by steveush** [comment]
@givemetraffic, sigh, it's posts like yours that make me wonder why I keep doing this. If you're not happy with it by all means say so and offer some constructive feedback but _"Sorry but, this is the worst fix ever."_ and _"Did you ever seen how other people implement autocomplete or filters?"_ are bordering on personal jabs at me. Please feel free to not use the plugin then if it's so terrible. _"Moreover it's default behavior now. Ridiculous."_ do you want to explain this or leave it as the rant it is like the rest of your reply?

Your proposed solution does not take into account that the input value gets set when the table is redrawn, it has to as if you use the API and apply a filter it must update the input with the current filter value. So simply moving the filtering outside of the table solves nothing.

Also I have tested with a 300ms delay and it all works fine apart from the caret being moved to the end of the input, @jeffconklin this is something I will look at fixing.

**Comment by givemetraffic** [comment]
@steveush No offense

>make me wonder why I keep doing this

Yeah. Me too. But if no one tell you that your code is sucks, how would u know that? It's not personal. It's about code.

>do you want to explain this

Sure. Destroying of input field in that user is typing, and recreating it after 100-200ms is ridiculous. It's mockery on poor users

> So simply moving the filtering outside of the table solves nothing.

You have a link to that $input object inside the instance, what is the problem?
I bet that this code will work further
```
if (search instanceof F.Filter){
  var query = search.query.val();
  if (this.exactMatch && this._exactRegExp.test(query)){
    query = query.replace(this._exactRegExp, '$1');
  }
  this.$input.val(query);
} else {
  this.$input.val(null);
}
```

**Comment by steveush** [comment]
@givemetraffic you still don't get it, I have no problem with issues being raised about the plugin and I do my best to resolve any problems. I've been doing this long enough to separate code from personal. It's the way you stated your issue, maybe English is not your first language and it's coming across more abrupt than you realize. I'm not asking for please and thank you, just some common decency and politeness.

Onto your above response, first of all there is no destroying and recreating of the input, it's the same one used from the first time the plugin is initialized.

There are two main issues, the first I have already stated, the input value is updated with each draw operation as filters can be applied through script and not using the UI at all.

The second issue is that the entire table is detached and then reattached to the DOM during it's draw operation. This was done to prevent a flash of unstyled content on first load and additional repaints with jQuery 3 in [this pull request](https://github.com/fooplugins/FooTable/pull/556). When the table is detached, the input automatically loses focus, therefore I have to manually reapply it. The changes I've made today ensure that the caret is placed back where it should be and users should be none the wiser.

If you're not happy with this again please do not use the plugin, or fork it and fix it yourself, all the code is available in this repo after all. Maybe even make a pull request seeing as you know how to fix it without causing other issues?

**Comment by givemetraffic** [comment]
@steveush oh man, detach() is destroying element, just with saving its state in a variable. DOM doesn't care about it, element is gone: btw second sentence in docs https://api.jquery.com/detach/

>you still don't get it

I get it. But don't expect people to fit your ideas about decency and politeness.

>If you're not happy with this again please do not use the plugin

We are not talking about my or yours preferences. I'm not using your plugin, I've just gave it a try.

Now I am trying to help people to have better experience. For both users and developers.

> This was done to prevent a flash of unstyled content on first load

So you are deleting entire table from DOM every time filter is called just because "unstyled content on first load". So sad. Sorry but fork or pull request won't help there

**Comment by steveush** [comment]
@givemetraffic thanks for the follow up post, it made me feel better knowing you don't understand the jQuery docs or how in memory objects work.

> oh man, detach() is destroying element, just with saving its state in a variable. DOM doesn't care about it, element is gone: btw second sentence in docs https://api.jquery.com/detach/

Removing an object from the DOM is not the same as destroying it, it remains in memory until it is no longer used. Here is some reading for you - [MDN - Memory Management](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Memory_Management). In fact jQuery's .detach() method was specifically created for this scenario, you use .remove() when the object is no longer required so jQuery can remove any data/event listeners bound to the element freeing it up to be garbage collected. When you use .detach() the actual DOM elements and there events are maintained so you can use them later. How do you think all the manipulations occur to the table while it is detached if it is destroyed? Magic?

> So you are deleting entire table from DOM every time filter is called just because "unstyled content on first load". So sad. Sorry but fork or pull request won't help there

As above, no the entire table is not being deleted, it is still in memory and being used. You also conveniently left off the second part of that sentence which is the main reason why it occurs on every draw operation, the fact that with jQuery 3, which uses requestAnimationFrame under the hood, was causing a lot of repaints in the browser.

> I get it. But don't expect people to fit your ideas about decency and politeness.

That's fine, I get it, you simply have none which is an excellent advertisement for you website and business.

> I'm not using your plugin, I've just gave it a try.

As you're not using the plugin and quite frankly don't know what you're talking about this will be my last response to you, I've already wasted enough time on you.

**Comment by givemetraffic** [comment]
@steveush 
>detach() is destroying element, just with saving its state in a variable

That means we will have a reference to variable, and all callbacks and event listeners, so garbage collector won't remove it from memory. That's what I am saying. You just didn't read.

Here is some reading for you - 
This is detach function https://github.com/jquery/jquery/blob/a6b0705294d336ae2f63f7276de0da1195495363/src/manipulation.js#L321

This is remove function that called from detach
https://github.com/jquery/jquery/blob/a6b0705294d336ae2f63f7276de0da1195495363/src/manipulation.js#L231

Did you see **unconditional** .removeChild() ?

Do you know what is .removeChild() ? Keep reading https://developer.mozilla.org/en-US/docs/Web/API/Node/removeChild

So I am absolutely sure now, that you are don't know what you are doing here.

Why you just not hide element while you "render something"? Why detach?

**Comment by givemetraffic** [comment]
Here is simple example for you http://codepen.io/givemetraffic/pen/bqzddJ
1) If you switch from detach() to hide()/show() you won't have issue with input focus and performance penalty from reinserting table from/to DOM
2) It's absolutely unnecessary to do this every time filter apply

**Comment by jeffconklin** [comment]
I would hope anything can be done to ensure focus is not lost at all. Though this plugin is not Google... what if the Google search bar.. or any search bar being used nowadays.. changed where your focus was while using the feature. @steveush ... if anything can be done to help this I would hope you may entertain it.

**Comment by jerfink** [comment]
@steveush,

Despite some rather ugly response text in this thread....
Thanks for adding the focus fix to the dev branch..  works like a charm.

**Comment by freerider7777** [comment]
steveush Thx for the useful library! Took dev branch - works great (with focus) :)

**Comment by peppeg85** [comment]
i'm sorry, but i still can't make it works...or better i did, but with the setInterval function and i don't like this workaround. as i can see the table is redrawn at every digitized letter, can i set the focus after the table is refreshed? where do you draw the table?

sorry for  my english and compliments for the plugin
